### PR TITLE
Improve compiler's checking for resetting an instantiated field

### LINKED ISSUE #16030: missing expected error when attempting to change LHS type in init=
See also PR #13904, PR #12428 and issue #12318.

It is my understanding that when an `init=` is called with a typed LHS that specifies a param or type field, it should be an error if the RHS type does not match at the point that the `init=` tries to set the already-set type/param field. Instead, it seems that the attempt to set the already-set type/param field is ignored.

``` chapel
record R {
  param fixed; //never allowed to change
  param p;
  var x: int;
}

proc R.init=(rhs: R) {
  this.fixed = rhs.fixed; // what happens if this.fixed is already false and rhs.fixed is true?

  // this one is set to the rhs if not provided already in the lhs
  this.p = if this.type.p != ? then this.type.p else rhs.p;

  this.x = rhs.x;
}

proc =(ref lhs:R, const ref rhs:R) {
  lhs.x = rhs.x;
}

proc main() {
  var start = new R(true, true, 1);
  writeln(start.type:string, " ", start);

  var b: R(false, false) = start; // init= -- expecting an error
  writeln(b.type:string, " ", b);
}
```

The program above compiles and runs (and indicates that the type of b is R(false, false) ).

Associated Future Test
[test/classes/initializers/initequals/changing-param-field-error3.chpl](https://github.com/chapel-lang/chapel/blob/704128ebc324df61722f086758c3d4fa951d596e/test/classes/initializers/initequals/changing-param-field-error3.chpl) (#16031)

### PR DESCRIPTION
**Pull request by mppf**
This PR adjusts `resolveInitField` in function resolution to check if a
generic field already provided a type is being reset.  This provides a
good error message in cases such as described in issue #16030. In order
to get that to work, I also needed to update
`AggregateType::fieldIsGeneric`  and `AggregateType::generateType` to
better handle the case of a field with a generic type and an
initialization expression.

This PR adds a better error for the case in which a record field is
declared with a generic type and an initialization expression. In that
event, there are open language design questions about how to default
initialize a variable of that type. See #16508.

While there, this PR removes the `decayToBorrow` argument and
functionality from several `resolveGenericActuals` functions as it is no
longer necessary now that we can pass generic type arguments. It also
replaces a strcmp on astrs with a pointer comparison.

Resolves #16030.
Helps with, but does not fully resolve, #16044.

Reviewed by @e-kayrakli - thanks!

- [x] primers pass with verify+valgrind and do not leak
- [x] full local futures testing

### CODE REVIEW

File: test/classes/initializers/initequals/changing-param-field-error3.good [hunk 1]
```diff
@@ -1 +1,4 @@
-a compilation error
+changing-param-field-error3.chpl:7: In method 'init=':
+changing-param-field-error3.chpl:8: error: Cannot replace an instantiated param field with another value
+changing-param-field-error3.chpl:8: note: field fixed has value false: bool but is set to true: bool
```
Comment (e-kayrakli) [inline test/classes/initializers/initequals/changing-param-field-error3.good:3]: can we put the values in a single quote or something?

File: test/classes/initializers/initequals/wrongType.good [hunk 2]
```diff
@@ -1,2 +1,4 @@
-wrongType.chpl:6: In function 'init=':
-wrongType.chpl:7: error: initializing field 'T' with type '2*int(64)', expecting 'int(64)'
+wrongType.chpl:6: In method 'init=':
+wrongType.chpl:7: error: Cannot replace an instantiated field with another type
+wrongType.chpl:7: note: field T has type int(64) but is set to 2*int(64)
```
Comment (e-kayrakli) [inline test/classes/initializers/initequals/wrongType.good:3]: I somehow feel more OK with types not being in quotes, but if you want to do that, that's also fine

**Review by e-kayrakli — APPROVED** [review]
Looks good!

I have one question, though: What is the behavior today when this doesn't happen inside any special function? By looking at the code changes I can't be sure whether this PR changes the behavior there.

```chapel
record R {
   param x: bool;
}

var r = new R(false);
r.x = true; // what is the error message here? 
```

I expect the message to be similar in that case. 

And can we add a test to lock that behavior in as well?

**Comment by mppf** [comment]
@e-kayrakli -

> I have one question, though: What is the behavior today when this doesn't happen inside any special function? By looking at the code changes I can't be sure whether this PR changes the behavior there.
> 
> ```chapel
> record R {
>    param x: bool;
> }
> 
> var r = new R(false);
> r.x = true; // what is the error message here? 
> ```
> 

It prints
```
bb.chpl:1: In module 'bb':
bb.chpl:6: error: illegal lvalue in assignment
bb.chpl:6: error: parameter set multiple times
```

This PR doesn't change the error message in this case. This PR only changes the errors when trying to replace such a field in an initializer.

> And can we add a test to lock that behavior in as well?

We already have tests along these lines such as `test/param/bharshbarg/setMultiple.chpl`.

**Comment by e-kayrakli** [comment]
Hmm that error message isn't bad either, but makes me wonder whether it can be made more verbose (as a future work, of course). The reason I am asking is that it may be less obvious for the user that the field `x` is a param, compared to `x` itself being a param.

**Pull request merged**